### ISSUE TITLE
[META] Use HttpClient Everywhere

### ISSUE THREAD

**Issue by shiftkey** [octokit/octokit.net#984]
One of my favourite bits of Octokit is how we take away all the hassles of configuring the HTTP stack, and for the most part it works. But every now and again, people hit up upon limitations with our setup - and you have to re-implement `HttpClientAdapter` to get to these features, and navigate a lot of the Octokit plumbing which isn't great.
## The Goals

Two main things come to mind:
- I want to move as much of the HTTP setup out of Octokit's core as possible, so that consumers have more control over the HTTP interactions
- I want to leverage the existing abstractions for `HttpClient` and `HttpMessageHandler` more broadly, so that Octokit can benefit from the extensibility in the .NET framework
## How to get there

As this is a mission that touches a lot of Octokit's core, I want to call out the tasks necessary - I'll open a milestone and break this down so others can help out along the way. I'd like to get this in before we call it "v1".
### The Right Abstraction

In #781 we settled on this sort of abstraction for defining the options available.

``` csharp
var info = new ClientInfo
{
    AppName = "my-cool-app",
    Credentials = new Credentials("my-token"),
    Server = "https://enterprise.my-work.com",
    UseDefaultProxy = true
};

var http = HttpClientFactory.Create(info);
var client = new GitHubClient(http);
```

We can probably support the existing `ctor`s for `GitHubClient` in the short term but eventually I'd like to see this go away, in favour of Just The Defaults™ - but we can argue about that in a PR:

``` csharp
var info = new ClientInfo
{
    AppName = "my-cool-app",
    Credentials = new Credentials("my-token"),
};

var client = new GitHubClient(info);
```
### Plumb The Depths

Once we've settled on that, and have proved out we can work this through to the core, the really hard stuff is to gut the internals to match this new abstraction. 
### The Experience

Once we've got the heavy lifting in place, we can revisit the current internals and see what improvements could be made.

I think we should keep `IConnection/IApiConnection` around for testability, but we should nudge them to be less public to consumers. Or fold them up into one. I'm not sure.
### Cleanup

There's a bunch of related cruft we can review and decide whether to drop once this is in place. I'll add more here as I work through laying the groundwork and remember what I was thinking about months ago when this started:
- `GitHubClient.Connection` has always been an escape hatch for being able to make HTTP requests with Octokit - that should now be unnecessary to keep around.
- Same thing with `GitHubClient.BaseAddress`

**Comment by Haacked** [comment]
A few comments.

```csharp
var info = new ClientInfo
{
    AppName = "my-cool-app",
    Credentials = new Credentials("my-token"),
    Server = "https://enterprise.my-work.com",
    UseDefaultProxy = true
};

var http = HttpClientFactory.Create(info);
```

One concern about this I forgot to bring up is that `http` here tends to be used as a singleton in many cases. And for the most part, that's fine. But `ClientInfo` is mutable here. Are we copying the values when we create the HTTP client?

Otherwise, we could run into subtle race conditions and concurrency issues if people change the values of the `ClientInfo` instance after the fact.

**Comment by Haacked** [comment]
Another comment, I think we need to enumerate a list of the type of settings people want globally vs per request to make sure this effort is worth doing as-is.

For example, as discussed in #985, `Timeout` to me seems like a per-request setting, though I'm OK with exposing a global `DefaultTimeout` property.

Another issue with the `ClientInfo` is by using properties to initialize it, we don't let the type system help people do the right thing by default.

For example, the reason we use `ProductHeaderValue` as a required `ctor` parameter today is because on the back end, the GitHub API requires a valid User Agent and getting the format for the user agent correct is actually trickier than one might expect.

Now, we could have overloads that let people specify an `AppName` and we create the `ProductHeaderValue` instead. I'm fine with that.

```csharp
var info = new ClientInfo(new ProductHeaderValue("my-cool-app", "1.0")
{
    Credentials = new Credentials("my-token"),
    Server = "https://enterprise.my-work.com",
    UseDefaultProxy = true
};

or

var info = new ClientInfo("my-cool-app")
{
    Credentials = new Credentials("my-token"),
    Server = "https://enterprise.my-work.com",
    UseDefaultProxy = true
};
```

**Comment by shiftkey** [comment]
@haacked 

> - But `ClientInfo` is mutable here. Are we copying the values when we create the HTTP client?
> - Another issue with the `ClientInfo` is by using properties to initialize it, we don't let the type system help people do the right thing by default.

Captured these as tasks over in #985 

> Now, we could have overloads that let people specify an `AppName` and we create the `ProductHeaderValue` instead. I'm fine with that.

I went down this path initially, but some tests we have have formats which I couldn't map to `ProductHeaderValue` correctly - likely due to using a version or comments. Thanks for opening #991 to dig into this.

**Comment by AArnott** [comment]
I'm just now starting to use Octokit.NET. I have a bunch of tests that verify my octokit code comes up with the right results, but that require live queries to github.com. The network dependency itself is unfortunate, but github also throttles down to 403's rather quickly such that we have to disable our tests.
I'd very much like to mock up the network traffic. I can do this if I could supply the `HttpMessageHandler` or `HttpClient` that the `GitHubClient` class uses, but I don't see any API to do that. Is this possible?  Is making that possible what _this_ issue is all about?

**Comment by ryangribble** [comment]
Hey @AArnott 

I'll need to read through this very old issue to completely grok what it was all about, but yes you can pass in your own `HttpClient` into one of the ctors for `GitHubClient` and then do whatever you want. You can also do like we do in our unit tests and pass in a mocked out `IConnection`... Or I guess you also could just mock out the `IGitHubClient` entirely. 

I guess it depends what you are trying to test, because if you go right down to the http client level you are really now testing octokit itself rather than just your code that's using octokit

**Comment by AArnott** [comment]
> but yes you can pass in your own HttpClient into one of the ctors for GitHubClient and then do whatever you want

If only that were true. Alas, no `GitHubClient` ctor accepts an `HttpClient` as a parameter.

Implementing `IConnection` looks *much* more laborious than implementing `HttpMessageHandler` because it has dozens of members instead of just one. And mocking up `IGitHubClient` assumes that I know how `GitHubClient` is *supposed* to work, but I don't. My tests should include executing actual `GitHubClient` code (and github.com server code, the first time) so that I verify I'm using the class correctly. If I mock up `IGitHubClient` then I can make it behave how I *think* it should behave rather than testing my code with how it *actually* behaves.

My strategy with other server-connected SDKs has been to inject my own `HttpMessageHandler` that *records* the operations the first time (while I'm writing the tests). Then I check in the recording so that all subsequent test runs can "play back" that network traffic and thus confirm that (assuming the server doesn't change behavior) my code continues to work.

**Comment by ryangribble** [comment]
Sorry the constructor I meant was on `Connection` and allows you to provide your own `HttpMessageHandler`

``` csharp
var connection = new Connection(new ProductHeaderValue("my-cool-app"),
    new HttpClientAdapter(() => myHttpMessageHandler));
var client = new GitHubClient(connection);
```

**Comment by ryangribble** [comment]
We were looking into record/replay type fixtures last year and @shiftkey  started a csharp implementation of "vcr" (a ruby record/replay based fixtures http fixtures library) and started looking at what we'd need to do in octokit to adapt to using it... 

you might find interesting: https://github.com/shiftkey/vcr-sharp

**Comment by AArnott** [comment]
Given vcr-sharp is abandoned, I started up a project of my own at https://github.com/aarnott/HttpClientEcho

It's not ready for use yet, but I hope to get it usable within a week.

**Comment by MaximRouiller** [comment]
Let me add you some details on why Octokit.NET for me is broken under Azure Functions.

I'm currently on a project using Azure Durable Functions. This runs under a consumption plan which itself is hosted within the [App Service Sandbox](https://github.com/projectkudu/kudu/wiki/Azure-Web-App-sandbox). 

This is interesting but what you're interested about in this massive page is called the [**Numerical Sandbox Limits**](https://github.com/projectkudu/kudu/wiki/Azure-Web-App-sandbox#numerical-sandbox-limits). 

![image](https://user-images.githubusercontent.com/209384/44154822-9f983f16-a079-11e8-9d32-583a98bcd826.png)

Everything in there is basically moot except one. Connections. It's limited at 300 per server on which it runs on. Azure Functions initially starts with one server executing all of your functions. Then it starts to scale up to 2, 3, N servers. Each of those "servers" have that 300 limit.

If you are creating a new `HttpClient` every time a request is made, [you end up running out of sockets](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/) as Windows waits 240 seconds to completely close the socket. That means that if I'm limited to a theoritical limit of 300 Octokit requests per servers per 4 minutes which translates to 1.25 requests per seconds. If I exceed that, I break the limit of the sandbox.

This literally makes Octokit unusable for Azure Functions on a consumption plan. 

This led me to go bare metal and use something like this: 

https://gist.github.com/MaximRouiller/74ae40aa994579393f52747e78f26441

```csharp
HttpClient client = new HttpClient();
client.BaseAddress = new Uri("https://api.github.com");
var token = "<token>";

client.DefaultRequestHeaders.UserAgent.Add(new System.Net.Http.Headers.ProductInfoHeaderValue("AppName", "1.0"));
client.DefaultRequestHeaders.Accept.Add(new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));
client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Token", token);

var response = await client.GetAsync("/user");
```

This allows me better control over the `HttpClient` and will allow me to survive the Connection sandbox limits.

If you have any questions, please fire away.

**Comment by khellang** [comment]
How is that different? You're still creating a new `HttpClient`, right?

**Comment by MaximRouiller** [comment]
I store it as static on my class.

**Comment by khellang** [comment]
No, it creates a `HttpClient` per `GitHubClient` instance. This is the code path when creating a `GitHubClient`:

https://github.com/octokit/octokit.net/blob/90b0a7c551e840312dd535a0ca285da5650f00ba/Octokit/GitHubClient.cs#L28-L29

https://github.com/octokit/octokit.net/blob/90b0a7c551e840312dd535a0ca285da5650f00ba/Octokit/Http/Connection.cs#L113-L114

https://github.com/octokit/octokit.net/blob/90b0a7c551e840312dd535a0ca285da5650f00ba/Octokit/Http/HttpClientAdapter.cs#L26-L35

And this actually sends the request:

https://github.com/octokit/octokit.net/blob/90b0a7c551e840312dd535a0ca285da5650f00ba/Octokit/Http/HttpClientAdapter.cs#L178-L184

As you can see; only one instance.

**Comment by shiftkey** [comment]
> As you can see; only one instance.

Oh cool, I legit thought something else was happening underneath. I thought @darrelmiller was scolding me about this at some stage but I can't find an issue about it.

> That means that if I'm limited to a theoritical limit of 300 Octokit requests per servers per 4 minutes which translates to 1.25 requests per seconds.

Keep in mind that Octokit will follow pagination links if a request has more results, so if you are using a `GetAll` API and the repository is non-trivial in terms of activity it's not a 1-1 mapping of request/response here.

>  If I exceed that, I break the limit of the sandbox.

What does "break" here entail? And how might we glean what the source of that "break" might be?

**Comment by khellang** [comment]
> that shouldn't mean new connections for each one.

I think that's exactly what it means, which is why [it's advised against](https://docs.microsoft.com/en-us/aspnet/web-api/overview/advanced/calling-a-web-api-from-a-net-client):

> `HttpClient` is intended to be instantiated once and reused throughout the life of an application.
> The following conditions can result in `SocketException` errors:
>
> - Creating a new `HttpClient` instance per request.
> - Server under heavy load.
>
> Creating a new `HttpClient` instance per request can exhaust the available sockets.

Related info:

https://blogs.msdn.microsoft.com/shacorn/2016/10/21/best-practices-for-using-httpclient-on-services/
https://medium.com/@nuno.caneco/c-httpclient-should-not-be-disposed-or-should-it-45d2a8f568bc
https://www.codeproject.com/Articles/1194406/Using-HttpClient-As-It-Was-Intended-Because-You-re
https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
http://www.nimaara.com/2016/11/01/beware-of-the-net-httpclient/

A static instance [can also be problematic](https://github.com/dotnet/corefx/issues/11224), which is why ASP.NET Core 2.1 introduced [HttpClientFactory](https://www.stevejgordon.co.uk/introduction-to-httpclientfactory-aspnetcore). There's also an ongoing issue about it [in corefx](https://github.com/dotnet/corefx/issues/21568).

**Comment by MaximRouiller** [comment]
@khellang I did get into issues. I had both errors from the sandbox itself and actual "permission exception" from the TCP stack rejecting new sockets opening.

That being confirmed, I'll check if there are other places where things might crash.

**Comment by ryangribble** [comment]
Were you using any IOC containers or anything that might cause multiple `GitHubClient` instances to be declared?

Generally whenever these types of things comes up around `HttpClient`, it does seem to be mostly solvable using the existing options provided in Octokit, and it is more just a lack of discoverability/documentation causing users to be unaware of them.  

The override on `Connection` that lets you specify an `IHttpClient` implementation means you could roll your own (admittedly a bit onerous given the number of methods to implement) or you can use the Octokit implementation of this interface (`HttpClientAdapter`) which lets you specify your own `HttpMessageHandler`.

In terms of Octokit's Http extensibility, I wonder if now .NET Core has the [new `HttpClientFactory`](https://blogs.msdn.microsoft.com/webdev/2018/02/28/asp-net-core-2-1-preview1-introducing-httpclient-factory/) stuff, we should consider reworking Octokit to use this (and allow the user to pass in their own)?

**Comment by MaximRouiller** [comment]
I don't think Octokit was the issue. I think it might be due internally in how Durable Functions works.

I have a workaround in place. I've already talked to the team to make sure we don't need a workaround. So either increasing limits of the sandbox or changing the Durable Functions defaults.

Let's see what they do.
